scipy/linalg/_matfuncs_sqrtm.c

```diff
@@ -0,0 +1,1006 @@
+static int sqrtm_recursion_z(SCIPY_Z* T, npy_intp bign, npy_intp n);
+static inline void zebra_pattern_s(float* restrict data, const Py_ssize_t n) {for (Py_ssize_t i=n-1; i>=0; i--) { data[2*i] = data[i]; data[2*i + 1] = 0.0f; }}
+static inline void zebra_pattern_d(double* restrict data, const Py_ssize_t n) {for (Py_ssize_t i=n-1; i>=0; i--) { data[2*i] = data[i]; data[2*i + 1] = 0.0; }}
+
+void
+matrix_squareroot_s(const PyArrayObject* ap_Am, float* restrict ret_data, int* isIllconditioned, int* isSingular, int* sq_info, int* view_as_complex)
+{
+    float aa, bb, cc, dd, cs, sn;
+    // Setting indicators to False.
+    // isComplex is set to True if the square-root turns out to be complex-valued (if exists)
+    // isIllconditioned is set to True if the solution of the Sylvester equation is ill-conditioned
+    // isSingular is set to True if the input is exactly singular
+    *view_as_complex = 0;
+    *isIllconditioned = 0;
+    *isSingular = 0;
+    int isComplex = 0;
+    int upcasted_to_complex = 0;
+
+    // --------------------------------------------------------------------
+    // Input Array Attributes
+    // --------------------------------------------------------------------
+    float* restrict Am_data = (float*)PyArray_DATA(ap_Am);
+    int ndim = PyArray_NDIM(ap_Am);              // Number of dimensions
+    npy_intp* shape = PyArray_SHAPE(ap_Am);      // Array shape
+    npy_intp n = shape[ndim - 1];                // Slice size
+    npy_intp* restrict strides = PyArray_STRIDES(ap_Am);
+    // Get the number of slices to traverse if more than one; np.prod(shape[:-2])
+    npy_intp outer_size = 1;
+    if (ndim > 2)
+    {
+        for (int i = 0; i < ndim - 2; i++) { outer_size *= shape[i];}
+    }
+
+    // --------------------------------------------------------------------
+    // Workspace computation and allocation
+    // --------------------------------------------------------------------
+    int info = 0, sdim = 0, lwork = -1, intn = (int)n;
+    float tmp_float = 0.0f, one = 1.0f, zero = 0.0f;
+    sgees_("V", "N", NULL, &intn, NULL, &intn, &sdim, NULL, NULL, NULL, &intn, &tmp_float, &lwork, NULL, &info);
+    // Improbable to fail at lwork query but check anyway
+    if (info != 0) { *sq_info = -100; }
+    lwork = (int)tmp_float;
+    // 2n*n + 2n*n for data and vs for potentially converting to complex if needed
+    // n + n for wr, wi (needed for gees calls)
+    // lwork for work
+    size_t buffer_size = 4*n*n + 2*n + lwork;
+    float* restrict buffer = malloc(buffer_size*sizeof(float));
+    if (buffer == NULL) { *sq_info = -101; return; }
+
+    // --------------------------------------------------------------------
+    // Pointers to the variables inside the buffer for easy access
+    // --------------------------------------------------------------------
+    float* restrict data = &buffer[0];
+    float* restrict data2 = &buffer[n*n];
```

Comment:
So you have a single buffer and partition it for various arrays and copies of the "inner" arrays.
What do you store at `data` and `data2`?

Reply:
I've added some comments above this.